### PR TITLE
feat(analytics-chart): support explore v4

### PR DESCRIPTION
**Pull request by filipgutica**
# Summary

BREAKING CHANGE: new chart data format

#### Resources

- [Creating a package docs](https://github.com/Kong/public-ui-components/blob/main/docs/creating-a-package.md)

### CODE REVIEW

File: packages/analytics/analytics-chart/src/composables/useExploreResultToDatasets.ts
Comment (adorack) [inline packages/analytics/analytics-chart/src/composables/useExploreResultToDatasets.ts:75]: I think you can get rid of the TODO?

File: packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts
Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts:315]: It's kind of a "better than nothing" situation for this test case. 

I'm using that pseudo mouse move event to trigger the tooltip, and it works pretty reliably, but there's no guarantee the tooltip will show up because of the nature of it, so the test does the check only if the tooltip actually shows up.

File: packages/analytics/analytics-chart/src/components/SimpleChart.cy.ts
Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/SimpleChart.cy.ts:28]: That's fine imo, it's just a component test, the component does't really care if it gets explore v4 dimensions or not, the structure is all that matters, this will still get rendered just fine

File: packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts
Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: The test was complaining, and was also not a reliable check. It was previously simply checking that the two arrays are arrays of length four, rather than actually testing the contents. So I turned it into a string for a better comparison

Comment (adamdehaven) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: Maybe I’m out of date on tests but what is `.eql()`?

Should this be `.eq()` or `.toEqual()`

File: packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts
Comment (adorack) [reply packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts:315]: That sounds like a recipe for a flaky test: if the subtitle ever doesn't exist, the test will sometimes pass (because the tooltip didn't show up) and sometimes fail (because subtitle wasn't there).  I think we should remove it entirely unless we have a way of guaranteeing that the tooltip shows up.

File: packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts
Comment (adorack) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: https://www.chaijs.com/api/bdd/#method_eql

> It was previously simply checking that the two arrays are arrays of length four, rather than actually testing the contents.

I don't see how that can be... and if that was how it was acting, then it's still just checking the length of the string...

Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: This is the error I get if I do it the old way....

image

Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: Even though, if you inspect, the contents of the array are the same 🤷

Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: The new way... 
image


Seems like a better check to me

Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: When I run the test on main, it still seems like it's just checking Array(4) against Array(4)? 

image

Comment (adamdehaven) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: > it's just checking Array(4) against Array(4)

Kind of related, I think you should stringify the arrays to compare them instead of comparing actual arrays; I'm not sure you're testing what you think you are.

```ts
expect(JSON.stringify(array1)).toEqual(JSON.stringify(array2))
```

Comment (adamdehaven) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: Sorry should have clarified -- at least if you stringify and it doesn't match, it will output the strings in the logs instead of `[ Array(4) ]`

Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/CsvExportModal.cy.ts:52]: Which is exactly what creating a string via joining is doing too 😄  

But I agree, using stringify is more elegant

File: packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts
Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts:315]: It will never produce false negatives which is what we don't want. 

The threshold that this produces false positives is low enough that I think it is still a valid test. 

If it fails, we know that there's an issue with the tooltip context. I'd like that to still be there, even if there's small chance this test will produce a false positive.

Comment (adorack) [reply packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts:315]: The thing I'm concerned about is that this creates a situation where the PR can randomly pass CI, but then fail once it's merged into main.  I think that's a situation that we'd like to avoid.

Is there any way to get rid of the conditional?  Can you calculate where the mouse needs to be in order to show the legend, then use a `cy.should` to wait for it to show up?

File: packages/analytics/analytics-chart/src/components/AnalyticsChart.vue
Comment (adorack) [inline packages/analytics/analytics-chart/src/components/AnalyticsChart.vue:241]: Minor: duplicate conditional.

File: packages/analytics/analytics-chart/src/components/TopNTable.vue
Comment (adorack) [inline packages/analytics/analytics-chart/src/components/TopNTable.vue:152]: Arrgh.  The cast here is necessary because `displayKey` is a string rather than a `QueryableExploreDimensions | ExploreAggregations`.  `displayKey` can be `''` in addition to a dimension or aggregation, so the cast is wrong, but TS doesn't seem to have any good way to represent "this value might not exist in the object", so perhaps we leave it alone for now.

File: packages/analytics/analytics-chart/src/locales/en.json
Comment (adorack) [inline packages/analytics/analytics-chart/src/locales/en.json:31]: I think you can get rid of the old translations (PascalCase and SCREAMING_SNAKE_CASE, below).  If they still surface under any circumstances, we should figure that out sooner rather than later.

File: packages/analytics/analytics-chart/src/locales/en.json
Comment (adorack) [inline packages/analytics/analytics-chart/src/locales/en.json:120]: Same comment about old metric names.

File: packages/analytics/dashboard-renderer/src/components/QueryDataProvider.vue
Comment (adorack) [inline packages/analytics/dashboard-renderer/src/components/QueryDataProvider.vue:81]: Yaaaaay!

**Review by adorack — CHANGES_REQUESTED** [review]

File: packages/analytics/analytics-chart/src/components/AnalyticsChart.vue
Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/AnalyticsChart.vue:241]: one is checking for `metric_names` and the other is checking for `metric_units`

File: packages/analytics/analytics-chart/src/components/TopNTable.vue
Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/TopNTable.vue:152]: Meaning just go back to have the key in DisplayBlob just be [key: string] ?

File: packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts
Comment (filipgutica) [reply packages/analytics/analytics-chart/src/components/AnalyticsChart.cy.ts]: Ok giving that a try, made some changes to the test

File: packages/analytics/analytics-chart/src/components/AnalyticsChart.vue
Comment (adorack) [reply packages/analytics/analytics-chart/src/components/AnalyticsChart.vue:241]: Aaah, sorry, my bad.

File: packages/analytics/analytics-chart/src/components/TopNTable.vue
Comment (adorack) [reply packages/analytics/analytics-chart/src/components/TopNTable.vue]: Ah, I'm sorry.  I actually wasn't asking for a change here, just trying to figure out whether there would be a better way in the future.  It's frustrating when TS can't quite figure out what line to walk between actual type safety and convenience.

**Review by adorack — DISMISSED** [review]
LGTM, though I'm also fine if you want to set DisplayBlob back to being keyed by aggregations / metrics.

**Review by adamdehaven — APPROVED** [review]
lgtm, also based on @adorack's approval

**Review by adorack — APPROVED** [review]

**Review by adamdehaven — APPROVED** [review]

**Pull request merged**
